### ISSUE TITLE
Enhance VRF export/import RT modeling

### ISSUE THREAD

**Issue by dmulyalin** [netbox-community/netbox#12093]
### NetBox version

v3.4.6

### Feature type

New functionality

### Proposed functionality

Add support to document Route-Target export and imports for VRF on a per-device basis by adding new VRF Export/Import model

### Use case

Documetn route-leaking between different VRF by adding new table that will link RT, VRF and devices together allowing to document which devices in which VRF export and import which RT.

Route leaking using route-targets is a common approach in networking to implement routing between VRF which is known as hard to keep track of. Adding new capability to Netbox will greatly enhance its offering on VRF documentation side.

Current models does not allow to capture information about what devices in which VRF import and export what route-targets. Current capabilities only allow to capture information on what RT imported and exported for given VRF, while in real world this always tied to actual network devices.

### Database changes

New model with this fields:
id
vrf - refers to existing VRF record
rt - refers to existing RT record
device - refers to existing device record
action - value of "import" or "export"

### External dependencies

Nill

**Comment by DanSheps** [comment]
You can already do this directly on VRF's.  You can add additional import targets onto a VRF.

Now, granted this would be difficult to use to properly generate configurations per device as some devices may import that additional route-target and some may not.  Perhaps it would make more sense to make the VRF model as a whole assigned per-device but that also introduces a number of difficulties.

**Comment by dmulyalin** [comment]
Adding route targets to vrf is great, but it does not capture which device exports or imports those route targets. Adding new table to hold those dependencies is what i hope could be the simplest way of doing it.

**Comment by DanSheps** [comment]
Yeah, I think the issue is here, with the way it is modelled in Netbox vs real life.

Like, you could have 3 devices, device A, device B, device C.  Lets say they are all ASR-1001-X's for simplicity.

On A you could have:

```
vrf def WAN
  rd 1:1
  address-family ipv4 unicast
    route-target both 1:1
    route-target import 1:5453
```

On B you could have:

```
vrf def B
  rd 1:5453
  address-family ipv4 unicast
    route-target both 1:5453
    route-target import 1:1
```

On C you could have:

```
vrf def B
  rd 1:5453
  address-family ipv4 unicast
    route-target both 1:5453
```

So like, on C you aren't importing 1:1 but on B you are.  It isn't possible to model the state of the vrf itself on each device currently in NetBox and you would have to use config contexts or intuit it currently.

I think this FR is proposing to more or less move the Route Target assignments to a device M2M model.  If I was doing this, I would do it this way:

Models:

```
VRF:
  name = string
  rd = string

RouteTarget:
  route-target = string

VRFDefinition:
  vrf = VRF
  route_target = RouteTarget
  device = Device

Interface:
  ...
  vrf = VRF
```

It would make the model slightly more complex but it would allow proper configuration generation.

**Comment by dmulyalin** [comment]
@DanSheps precisely yes, this is the use case I was trying to document in Netbox but apart from using config contexts or change vrf to per-device by encoding device name in vrf name, was not able to come up with any other workarounds.

@jeremystretch assigning interfaces to VRF is great and very representative of how real networks work. Current VRF and RT modeling approach, however, does not capture the relationship between RT and VRF on a per-device basis, which is how it is done in the networks. In fact, VRF name is having device local significance, and (as you of course aware) RT control the routing between VRFs on different devices. 

I was hoping that addint new `vpn_import_export` model would be able to serve us to capture RT usage for VRFs on a per-device basis. At the end it would be good to end up with this table that user can interface with:
| id  | vrf  | device  | rt  | action   | description  |
|---|---|---|---|---|---|
| 1  | FOO  | router-1  | 1:123  | import  | Import spoke routes  |
| 2  |  BAR | router-2  | 1:234  | export  | Export spoke routes  |
| 3  | FOOBAR  | router-3  | 1:234  | both  | HUB, receive all routes from all spokes  |

Same `vpn_import_export` can be used for L2VPN as well, or any other future entities that rely on route-targets for NLRI exchange. Moreover, having dedicated `vpn_import_export` model likely would allow to remove relationship between RT and VRF as it is not fully representative of how real networks implemented.

**Comment by DetunizedGravity** [comment]
> Yeah, I think the issue is here, with the way it is modelled in Netbox vs real life.
> 
> Like, you could have 3 devices, device A, device B, device C. Lets say they are all ASR-1001-X's for simplicity.
> 
> On A you could have:
> 
> ```
> vrf def WAN
>   rd 1:1
>   address-family ipv4 unicast
>     route-target both 1:1
>     route-target import 1:5453
> ```
> 
> On B you could have:
> 
> ```
> vrf def B
>   rd 1:5453
>   address-family ipv4 unicast
>     route-target both 1:5453
>     route-target import 1:1
> ```
> 
> On C you could have:
> 
> ```
> vrf def B
>   rd 1:5453
>   address-family ipv4 unicast
>     route-target both 1:5453
> ```
> 
> So like, on C you aren't importing 1:1 but on B you are. It isn't possible to model the state of the vrf itself on each device currently in NetBox and you would have to use config contexts or intuit it currently.
> 
> I think this FR is proposing to more or less move the Route Target assignments to a device M2M model. If I was doing this, I would do it this way:
> 
> Models:
> 
> ```
> VRF:
>   name = string
>   rd = string
> 
> RouteTarget:
>   route-target = string
> 
> VRFDefinition:
>   vrf = VRF
>   route_targets = [RouteTarget]
>   device = [Device]
> 
> Interface:
>   ...
>   vrf = VRF
> ```
> 
> It would make the model slightly more complex but it would allow proper configuration generation.

I would go as far as to say that if one wants to truly describe the real world, vrf names need not be unique, since they are local to a device. The actual primary key for VRF should be (device id, name). Dunno if Netbox and its foundations allow that, though.

**Comment by davidgwatkins** [comment]
Stumbled across this issue searching for per-device route distinguisher support.  I really like `VRFDefinition` model idea!  Would propose also moving the VRF RD into that model to allow unique RDs per PE (Type 1 RD per RFC 4364).

Another argument for more granular RT control: CSPs often reuse one VRF globally with multiple regional routing domains and associated ASNs / RTs.  When I modeled this in the past, I used a Custom Field on the VRF model with an assigned number field and varied the admin fields for both RD and RT using Region-based Config Contexts.  The `VRFDefinition` model would eliminate the need for that workaround and ambiguity.

**Comment by DanSheps** [comment]
I came across this searching for something else.  Let me model out the ERD in mermaid I would be willing to take this on in 4.0 or later.

**Comment by DanSheps** [comment]
[![](https://mermaid.ink/img/pako:eNqtlM1uwjAMx1_Fyhl4gJ4R0oQ2IYZ26iUiLrVGk8px0RDtuy9pi1oV9sG0XJK6v_jj7yQXtXcGVaKQl6QPrIvUQhhLPNEe4dJ9xUFWgAxs1pCqDVOh-QxrPKdqQLww2QNYXSAE6iXOLut9jcHoy2BGloSc9bCKTt-2q4BejcB41HFx3dd0U6T-Javg6A7FJjJbVwmCIS_BRj5HnmTRAjvNB5S_ZcPRwVw6D4He5dg5hd5GBq1QRjehQ94jlR4LHpETZ53e215gn1MJ4qaCdD1qT8Fd_LarXws96uukmicryJl--KhdK_m-jhBkODW1WyzcZaJfArn294St3Xwe6HGjE6CidCy_5_FjxPdXqvkhjR6r6xYb9Bkhg7GNWrfl9f9Tq2aqQC40mXCtW1VTJTkWoVdJWBrN71GfJnC6Evd6tnuVCFc4U1VptGD_EKgk00cfrGhIHD9370T7XDSfNURSoA?type=png)](https://mermaid.live/edit#pako:eNqtlM1uwjAMx1_Fyhl4gJ4R0oQ2IYZ26iUiLrVGk8px0RDtuy9pi1oV9sG0XJK6v_jj7yQXtXcGVaKQl6QPrIvUQhhLPNEe4dJ9xUFWgAxs1pCqDVOh-QxrPKdqQLww2QNYXSAE6iXOLut9jcHoy2BGloSc9bCKTt-2q4BejcB41HFx3dd0U6T-Javg6A7FJjJbVwmCIS_BRj5HnmTRAjvNB5S_ZcPRwVw6D4He5dg5hd5GBq1QRjehQ94jlR4LHpETZ53e215gn1MJ4qaCdD1qT8Fd_LarXws96uukmicryJl--KhdK_m-jhBkODW1WyzcZaJfArn294St3Xwe6HGjE6CidCy_5_FjxPdXqvkhjR6r6xYb9Bkhg7GNWrfl9f9Tq2aqQC40mXCtW1VTJTkWoVdJWBrN71GfJnC6Evd6tnuVCFc4U1VptGD_EKgk00cfrGhIHD9370T7XDSfNURSoA)

**Comment by dmulyalin** [comment]
Am i reading right that vrfdefinition gonna be tied to device? If so, rd could be moved to vrfdefinition model as well, as rd always of device local significance e.g. ot uncommon rd could encode device loopback ip to be device specific.

**Comment by dmulyalin** [comment]
Also think interface relationship with vrf need to be with vrfdefinition model if vrfdefinition represents an instance of the vrf. In that case the original vrf model does not needed? Could it be rebuild as say l3domain model, representing routing contingency across multiple vrfs?

**Comment by DanSheps** [comment]
> Also think interface relationship with vrf need to be with vrfdefinition model if vrfdefinition represents an instance of the vrf. In that case the original vrf model does not needed? Could it be rebuild as say l3domain model, representing routing contingency across multiple vrfs?

VRF is tied to other things within netbox beyond those modelled here.  This is a simple snippet of the relevant relationships for this specific FR.

**Comment by ajackson79** [comment]
Adding my voice to the list. We use a unique RD per device so routes are unique across the network. We would like to model that in Netbox. Having device specific RD and RTs would be ideal for flexibility.

**Comment by DanSheps** [comment]
I think this might be a good candidate for 4.2

**Comment by ajackson79** [comment]
> I think this might be a good candidate for 4.2

Yes I agree this could be a great candidate for 4.2 and the SP focus. We would then not have to use config contexts to rewrite RD and RT per devices.
